### PR TITLE
Adding interfaces and final , static methods, class and variable

### PR DESCRIPTION
**Pull request by asurelaba**
Homework 4.
This pull request includes following changes:

1. Add 5 interfaces to the existing hierarchy. - ISupplyChain, IFileTaxes, Isalary, IWorkingShift, IFoodSafetyChecks, IMaintainColdSection
2. Use polymorphism with the abstract class and interface from the hierarchy. - Using Item class variable that can also access it child class methods. 
3. Create final class, method, variable. - Child classes for whom further inheritance is not required. 
4. Create a static block, method, variable.

### CODE REVIEW

File: entities/BillingCounter.java [hunk 1]
```diff
@@ -43,18 +43,32 @@ public void setEmployee(Employee employee) {
...
         return receipts.size();
     }
+
+    public static float calculatePrice(float weight, float pricePerPound) {
+        return weight * pricePerPound;
+    }
+
+    public static float getWeight(Item item) {
+        return 1.0f;
+    }
```
Comment (ankochergin) [inline entities/BillingCounter.java]: item is unused

File: entities/FreezerAsile.java [hunk 2]
```diff
@@ -21,4 +29,50 @@ public float getTemp() {
...
         this.temp = temp;
     }
+
+    public Date getLastCleaned() {
+        return lastCleaned;
+    }
+
+    public void setLastCleaned(Date lastCleaned) {
+        this.lastCleaned = lastCleaned;
+    }
```
Comment (ankochergin) [inline entities/FreezerAsile.java]:
```suggestion
    public Date getLastCleanedDate() {
        return lastCleaned;
    }

    public void setLastCleanedDate(Date lastCleaned) {
        this.lastCleaned = lastCleaned;
    }
```

File: entities/IMaintainColdSection.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+    float freezerMinTemp = 32;
+    float refrigiratorMinTemp = 40;
+
+    boolean isColdStorageCleaned();
+
+    boolean hasBackupPower();
+
+    void maintainTemp();
+}
```
Comment (ankochergin) [inline entities/IMaintainColdSection.java]: be consistent in formatting, please check where you have empty lines in other interfaces

File: entities/BillingCounter.java (hunk 1 above)
Comment (asurelaba) [reply entities/BillingCounter.java]: Will make the changes.

File: entities/Employee.java [hunk 4]
```diff
@@ -72,10 +76,24 @@ public void setManager(Manager manager) {
...
+        System.out.println("Employee Id:" + employeeId + "\n Name: " + name + "\nAddress: " + address + "\nPhone: " + phone + "\n Role: " + role);
         if (manager != null) {
             System.out.println("Manager: " + manager);
         }
     }
+
+    @Override
+    public void salaryHike(float percentage) {
+        salary += (salary * percentage);
+    }
```
Comment (ankochergin) [inline entities/Employee.java]: I think only manager can increase employees salary. Maybe we should move this to Manager and make method salaryHike(employee, percentage)? what do you think?

File: entities/FreezerAsile.java [hunk 5]
```diff
@@ -21,4 +29,50 @@ public float getTemp() {
...
+
+    @Override
+    public void removePerishableBeforeExpiry() {
+        PerishableItem perishableItem;
+        int i = 0;
+        while (i < itemsInShelf.size()) {
+            // System.out.println("in refrigirator" + itemsInShelf.get(i).getItemName() + LocalDate.now());
+            perishableItem = (PerishableItem) itemsInShelf.get(i);
+            if (LocalDate.now().isAfter(perishableItem.bestBefore)) {
+                //System.out.println("Removing" + perishableItem.getItemName() + " from refrigirator" + LocalDate.now().compareTo(perishableItem.bestBefore));
```
Comment (ankochergin) [inline entities/FreezerAsile.java]: uncomment lines or remove them if unused

File: entities/Shop.java [hunk 6]
```diff
@@ -158,16 +173,61 @@ public void placeOrder() {
...
+        //ask inventory to check through all items for expiry
+
+        for (Asile asile : asiles) {
+            if (asile.getClass() == FreezerAsile.class) {
+                ((FreezerAsile) asile).removePerishableBeforeExpiry();
+            } else if (asile.getClass() == RefrigiratorAsile.class) {
+                ((RefrigiratorAsile) asile).removePerishableBeforeExpiry();
+            }
+        }
+    }
```
Comment (ankochergin) [inline entities/Shop.java]: documentation (comments what methods do) should be above methods

File: entities/Employee.java (hunk 4 above)
Comment (asurelaba) [reply entities/Employee.java]: will make changes

File: entities/FreezerAsile.java (hunk 5 above)
Comment (asurelaba) [reply entities/FreezerAsile.java]: The comments are removed.

File: entities/BillingCounter.java (hunk 1 above)
Comment (asurelaba) [reply entities/BillingCounter.java]: This method is modified to:
public static float getWeightOfItemInCart(Item item) {
        Random random = new Random();
        if (item.getClass() == FreshProduceItem.class) {
            return random.nextFloat(10);
        }
        return 1;
    }

File: entities/Employee.java (hunk 4 above)
Comment (asurelaba) [reply entities/Employee.java]: salaryHike is moved to Manager class. Shop calls the manager to increase the salary by a certain percentage.

**Review by ankochergin — APPROVED** [review]

**Pull request merged**